homeassistant/components/qbus/config_flow.py

```diff
@@ -0,0 +1,158 @@
+            _LOGGER.debug("Payload empty")
+            return self.async_abort(reason="invalid_discovery_info")
+
+        match discovery_info.subscribed_topic:
+            case self._gateway_topic:
+                return await self._async_handle_gateway_topic(discovery_info)
+
+            case self._config_topic:
+                return await self._async_handle_config_topic(discovery_info)
+
+            case self._device_topic:
+                return await self._async_handle_device_topic(discovery_info)
+
+        return self.async_abort(reason="invalid_discovery_info")
+
+    async def async_step_discovery_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Confirm the setup."""
+        if TYPE_CHECKING:
+            assert self._device is not None
+
+        if user_input is not None:
+            return self.async_create_entry(
+                title=f"Controller {self._device.serial_number}",
+                data={
+                    CONF_SERIAL: self._device.serial_number,
+                    CONF_ID: self._device.id,
+                },
+            )
+
+        return self.async_show_form(
+            step_id="discovery_confirm",
+            description_placeholders={
+                CONF_SERIAL: self._device.serial_number,
+            },
+        )
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initialized by the user."""
+        return self.async_abort(reason="not_supported")
+
+    async def _async_handle_gateway_topic(
+        self, discovery_info: MqttServiceInfo
+    ) -> ConfigFlowResult:
+        _LOGGER.debug("Handling gateway state")
+        gateway_state = self._message_factory.parse_gateway_state(
+            discovery_info.payload
+        )
+
+        if gateway_state is not None and gateway_state.online is True:
+            _LOGGER.debug("Requesting config")
+            await mqtt.async_publish(
+                self.hass, self._topic_factory.get_get_config_topic(), b""
+            )
+
+        # Abort to wait for config topic
+        return self.async_abort(reason="invalid_discovery_info")
+
+    async def _async_handle_config_topic(
+        self, discovery_info: MqttServiceInfo
+    ) -> ConfigFlowResult:
+        _LOGGER.debug("Handling config topic")
+        qbus_config = self._message_factory.parse_discovery(discovery_info.payload)
+
+        if qbus_config is not None:
+            QbusConfigContainer.store_config(self.hass, qbus_config)
+
+            _LOGGER.debug("Requesting device states")
+            device_ids = [x.id for x in qbus_config.devices]
+            request = self._message_factory.create_state_request(device_ids)
+            await mqtt.async_publish(self.hass, request.topic, request.payload)
+
+        # Abort to wait for device topic
+        return self.async_abort(reason="invalid_discovery_info")
```

Comment:
Could we have a better reason in these aborts, or are we required to use this one?